SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -81,72 +142,240 @@ abbrev R := (ZMod q)[X] ⧸ (Ideal.span {f q n})
+  NOTE: Lean times out if I use the abbreviation `R.fromPoly` for unclear reasons! -/
+theorem R.fromPoly_representatitive' (a : R q n) : R.fromPoly (R.representative' q n a) = ↑ a := by
+  simp[R.fromPoly, R.representative']
+  apply Function.surjInv_eq
 
-noncomputable instance {q n : Nat} : Zero (R q n) := ⟨R.zero⟩
-noncomputable instance {q n : Nat} : One (R q n) := ⟨R.one⟩
+theorem R.fromPoly_representatitive'_toFun (a : R q n) : (R.fromPoly (q := q) (n := n)).toFun (R.representative' q n a) = ↑a := by
+  apply Function.surjInv_eq
 
-private noncomputable def R.representative' : R q n → (ZMod q)[X] := Function.surjInv Ideal.Quotient.mk_surjective
 /--
 The representative of `a : R q n` is the (unique) polynomial `p : ZMod q[X]` of degree `< 2^n`
  such that `R.fromPoly p = a`.
 -/
 noncomputable def R.representative : R q n → (ZMod q)[X] := fun x => R.representative' q n x %ₘ (f q n)
 
+@[simp]
+theorem R.fromPoly_kernel_eq_zero (x : (ZMod q)[X]) : R.fromPoly (n := n) (f q n * x) = 0 := by
+   unfold fromPoly
+   apply Ideal.Quotient.eq_zero_iff_mem.2
+   rw [Ideal.mem_span_singleton]
+   simp [Dvd.dvd]
+   use x
+
 /--
-`R.representative` is in fact a representative of the equivalence class.
+`R.'representative'` is in fact a representative of the equivalence class.
 -/
 @[simp]
 theorem R.fromPoly_representative : forall a : R q n, (R.fromPoly (n:=n) (R.representative q n a)) = a := by
  intro a
  simp [R.representative]
  rw [Polynomial.modByMonic_eq_sub_mul_div _ (f_monic q n)]
  rw [RingHom.map_sub (R.fromPoly (q := q) (n:=n)) _ _]
- have hker : forall x, fromPoly (f q n * x) = 0 := by
-   intro x
-   unfold fromPoly
-   apply Ideal.Quotient.eq_zero_iff_mem.2
-   rw [Ideal.mem_span_singleton]
-   simp [Dvd.dvd]
-   use x
- rw [hker _]
+ rw [R.fromPoly_kernel_eq_zero]
  simp
  apply Function.surjInv_eq
 
 
 /--
-Characterization theorem for any potential representative.
+Characterization theorem for any potential representative (in terms of ideals).
 For an  `a : (ZMod q)[X]`, the representative of its equivalence class
 is just `a + i` for some `i ∈ (Ideal.span {f q n})`.
 -/
-theorem R.fromPoly_rep'_eq : forall a : (ZMod q)[X], ∃ i ∈ Ideal.span {f q n}, (R.fromPoly (n:=n) a).representative' = a + i := by
+theorem R.fromPoly_rep'_eq_ideal : forall a : (ZMod q)[X], ∃ i ∈ Ideal.span {f q n}, (R.fromPoly (n:=n) a).representative' = a + i := by
   intro a
   exists (R.fromPoly (n:=n) a).representative' - a
   constructor
   · apply Ideal.Quotient.eq.1
     simp [R.representative', Function.surjInv_eq]
   · ring
   done
+/--
+Characterization theorem for any potential representative (in terms of elements).
+For an  `a : (ZMod q)[X]`, the representative of its equivalence class
+is a concrete element of the form `a + k * (f q n)` for some `k ∈ (ZMod q)[X]`.
+-/
+theorem R.fromPoly_rep'_eq_element (a : (ZMod q)[X]) : ∃ (k : (ZMod q)[X]), (R.fromPoly (n:=n) a).representative' = k * (f q n) + a := by
```

Comment:
Try to just read the statement to name the theorem. `exists_representative_fromPoly_eq_mul_add`

Reply:
But if you used `mod`, then `k` would just be `div` right. And it's nicer to get rid of the exists.